subsys/nrf_rpc/nrf_rpc_rpmsg.c

```diff
@@ -85,45 +91,86 @@ static void ll_event_handler(struct rp_ll_endpoint *endpoint,
 	receive_callback(buf, length);
 }
 
-int nrf_rpc_tr_init(nrf_rpc_tr_receive_handler_t callback)
+static int endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
+	uint32_t src, void *priv)
 {
-	int err = 0;
+	if (len == 0) {
+		if (!is_handshake_done) {
+#if !MASTER
+			k_sem_take(&remote_init_sem, K_FOREVER);
+#endif
+			rpmsg_send(ept, (uint8_t *)"", 0);
+
+			is_handshake_done = true;
+			LOG_INF("Handshake done");
+			event_handler(NRF_RPC_EVENT_CONNECTED, NULL, 0);
+		}
+		return RPMSG_SUCCESS;
+	}
 
-	NRF_RPC_ASSERT(callback != NULL);
+	event_handler(NRF_RPC_EVENT_DATA, data, len);
 
-	receive_callback = callback;
+	return RPMSG_SUCCESS;
+}
 
-	err = rp_ll_init();
-	if (err != 0) {
-		goto error_exit;
-	}
+static int nrf_rpc_register_endpoint(const struct device *dev)
+{
+	ARG_UNUSED(dev);
 
-	err = rp_ll_endpoint_init(&ll_endpoint, 1, ll_event_handler, NULL);
-	if (err != 0) {
-		goto error_exit;
-	}
+	int err = rpmsg_service_register_endpoint("nrf_rpc", endpoint_cb);
 
-	k_sem_take(&handshake_sem, K_FOREVER);
+	endpoint_id = err;
 
-	NRF_RPC_DBG("nRF RPC Initialized");
+	if (err < 0) {
+		LOG_ERR("Registering endpoint failed with %d", err);
+		return err;
+	}
 
 	return 0;
+}
 
-error_exit:
-	NRF_RPC_DBG("nRF RPC Initialization error %d", err);
+int nrf_rpc_tr_init(nrf_rpc_tr_receive_handler_t callback)
+{
+	NRF_RPC_ASSERT(callback != NULL);
+	receive_callback = callback;
 
-	return translate_error(err);
+#if !MASTER
+	NRF_RPC_INF("RPC remote");
+	k_sem_give(&remote_init_sem);
+#endif
+
+#if MASTER
+	NRF_RPC_INF("RPC master");
+	while (!rpmsg_service_endpoint_is_bound(endpoint_id)) {
+		k_sleep(K_MSEC(1));
+	}
+
+	NRF_RPC_DBG("RPC master sem take");
+	rpmsg_service_send(endpoint_id, (uint8_t *)"", 0);
+	k_sem_take(&master_handshake_sem, K_FOREVER);
+#endif
```

Comment:
You can use #if #else construct for this and put the RPC master block first as it easier to read not negated definitions.

Reply:
Updated following the advice.